File: examples/swift/micro/Gateway.swift
import Foundation

class Gateway {
    let userService = UserService()

    func handleRequest(queryParams: [String: String]) -> String {
        let token = queryParams["token"]!    // source: user input
        let action = queryParams["action"]!  // source: user input

        let user = userService.getUser(token: token)         // flows to SQL injection
        let result = userService.updateUser(token: token, action: action) // flows to command injection

        return "{\"user\": \"\(user.id ?? "null")\", \"result\": \"\(result ?? "null")\"}"
    }
}

File: examples/cpp/micro/gateway.cpp
#include "gateway.h"
#include "user_service.h"
#include <iostream>

namespace micro {

std::string handle_request(const std::map<std::string, std::string>& params) {
    std::string token = params.at("token");    // source: user input
    std::string action = params.at("action");  // source: user input

    UserInfo user = get_user(token);           // flows to SQL injection
    std::string result = update_user(token, action); // flows to command injection

    return "{\"user\": \"" + user.id + "\", \"result\": \"" + result + "\"}";
}

}
